test/test_EKF_terrain_estimator.cpp

```diff
@@ -0,0 +1,176 @@
+		_sensor_simulator._gps.setPositionRate(simulated_velocity);
+
+		// Configure range finder simulator data
+		_sensor_simulator._rng.setData(rng_height, 100);
+		_sensor_simulator._rng.setLimits(0.1f, 20.f);
+		_sensor_simulator.startRangeFinder();
+
+		// Configure optical flow simulator data
+		flowSample flow_sample = _sensor_simulator._flow.dataAtRest();
+		flow_sample.flow_xy_rad =
+			Vector2f( simulated_velocity(1) * flow_sample.dt / flow_height,
+				 -simulated_velocity(0) * flow_sample.dt / flow_height);
+		_sensor_simulator._flow.setData(flow_sample);
+		const float max_flow_rate = 5.f;
+		const float min_ground_distance = 0.f;
+		const float max_ground_distance = 50.f;
+		_ekf->set_optical_flow_limits(max_flow_rate, min_ground_distance, max_ground_distance);
+		_sensor_simulator.startFlow();
+
+		_ekf->set_in_air_status(true);
+		_sensor_simulator.runSeconds(8);
+	}
+};
+
+TEST_F(EkfTerrainTest, setFlowAndRangeTerrainFusion)
+{
+	// WHEN: simulate being 5m above ground
+	// By default, both rng and flow aiding are active
+	const float simulated_distance_to_ground = 1.f;
+	_sensor_simulator._rng.setData(simulated_distance_to_ground, 100);
+	_sensor_simulator._rng.setLimits(0.1f, 9.f);
+	_sensor_simulator.startRangeFinder();
+	_ekf->set_in_air_status(true);
+	_sensor_simulator.runSeconds(1.5f);
+
+	// THEN: By default, both rng and flow aiding are active
+	EXPECT_TRUE(_ekf_wrapper.isIntendingTerrainRngFusion());
+	EXPECT_TRUE(_ekf_wrapper.isIntendingTerrainFlowFusion());
+	const float estimated_distance_to_ground = _ekf->getTerrainVertPos();
+	EXPECT_FLOAT_EQ(estimated_distance_to_ground, simulated_distance_to_ground);
+
+	// WHEN: rng fusion is disabled
+	_ekf_wrapper.disableTerrainRngFusion();
+	_sensor_simulator.runSeconds(0.2);
+
+	// THEN: only rng fusion should be disabled
+	EXPECT_FALSE(_ekf_wrapper.isIntendingTerrainRngFusion());
+	EXPECT_TRUE(_ekf_wrapper.isIntendingTerrainFlowFusion());
+
+	// WHEN: flow is now diabled
+	_ekf_wrapper.disableTerrainFlowFusion();
+	_sensor_simulator.runSeconds(0.2);
+
+	// THEN: flow is now also disabled
+	EXPECT_FALSE(_ekf_wrapper.isIntendingTerrainRngFusion());
+	EXPECT_FALSE(_ekf_wrapper.isIntendingTerrainFlowFusion());
+}
+
+TEST_F(EkfTerrainTest, testFlowForTerrainFusion)
+{
+	// GIVEN: flow for terrain enabled but not range finder
+	_ekf_wrapper.enableTerrainFlowFusion();
+	_ekf_wrapper.disableTerrainRngFusion();
+
+	// WHEN: the sensors do not agree
+	const float rng_height = 1.f;
+	const float flow_height = 5.f;
+	runFlowAndRngScenario(rng_height, flow_height);
+
+	// THEN: the estimator should use flow for terrain and the estimated terrain height
+	// should converge to the simulated height
+	EXPECT_FALSE(_ekf_wrapper.isIntendingTerrainRngFusion());
+	EXPECT_TRUE(_ekf_wrapper.isIntendingTerrainFlowFusion());
+
+	const float estimated_distance_to_ground = _ekf->getTerrainVertPos();
+	EXPECT_NEAR(estimated_distance_to_ground, flow_height, 0.5f);
```

Comment:
Do you have an idea why the error is here higher?

Reply:
I think this is because the range finder fusion is stronger than the optical flow one. If we let it run more, it will converge to the true value.